### ISSUE TITLE
Video Corruption Showing on FireTVs

### ISSUE THREAD

**Issue by SAPierce2006** [google/ExoPlayer#9367]
We have a production issue where our users started complaining about certain videos playing artifacted / corrupted on FireTVs, which is the primary device our customers use. Example videos are at the bottom.

* We're using h.264 High Profile Level 4 for our encoding.
* Not all videos are effected. Only certain videos in their video loop are impacted. Once a video is affected, it seems to stay affected, even past app restarts.
* A user sent us one of the devices they are seeing this on. I was able to reproduce it for a few hours, and then it magically resolved itself.
* We are storing the videos encrypted, and using exoplayer to decrypt them at runtime.

I've exhausted all options I have, and am no longer sure where to look for a solution for this. More and more customers seem to reach out every week with this issue. I realize you probably won't be able to just fix this because I'm not able provide much information. 

**I'm unsure of how to proceed in debugging this. I'll do anything you ask me to, or get any information necessary. Thank you.**

https://user-images.githubusercontent.com/89875574/131575266-3ccb694f-acbb-4ae5-9826-bcf51a0dc7a9.mov

https://user-images.githubusercontent.com/89875574/131575268-828490ea-8cd6-485e-8b6b-da0890610e42.mp4

**Comment by kim-vde** [comment]
The error could have multiple sources:
- A bug in the decoder of the FireTV. Have you tried to play the problematic contents on other devices? If it works, then this might be the problem.
- The input stream that is invalid. Does the same content plays on other players (like VLC or ffplay for instance)? Do you see codec errors or warning in the logs indicating that the content is invalid?
- There is a bug in ExoPlayer. To investigate this, we would need a way to reproduce I am afraid.

**Comment by SAPierce2006** [comment]
@kim-vde Thanks for responding!

> Have you tried to play the problematic contents on other devices?

The content plays fine on all devices, including the FireTVs that that it messes up on, but seemingly out of nowhere the device will start showing this error on a single video in a video loop of 8-15 videos, and the corruption will be there EVERY time the video plays.

It plays fine locally on other FireTVs and other AndroidTVs. It plays locally correctly on our computers on VLC etc.

The error persists across app restarts and device restarts. I had a customer ship us one of the devices, and I was able to reproduce it, but only in the production application. Internal version of our app were unable to reproduce it on the same device. The next day, the issue resolved itself on the production version and I couldn't reproduce it anymore. Other customers continue to have the issue though.

> Does the same content plays on other players (like VLC or ffplay for instance)?

Yes.

> Do you see codec errors or warning in the logs indicating that the content is invalid?

We haven't been able to reproduce it locally on an internal build so that I can see the logs. 

1. Are the logs you are talking about output directly to logcat? Is there any way we could intercept them so that we could save and view them remotely?
2. Out of curiosity, do you have recommended ffmpeg settings to achieve maximum compatibility?

**Comment by SAPierce2006** [comment]
I was able to reproduce it locally just now. The static was on one video, and then it couldn't be reproduced again, by me rewinding to the previous video. I was plugged in and was able to capture the logs. The FireTV LogCat is small, and I wasn't able to get the beginning of the logs I don't think.

I also took logs from the same FireTV when it's playing a video correctly, and there are no errors, for comparison. 

**Does anything look odd to you in these logs?**

[Logs-VideoWithStaticError.txt](https://github.com/google/ExoPlayer/files/7101216/Logs-VideoWithStaticError.txt)
[Logs-VideoPlayingCorrectly.txt](https://github.com/google/ExoPlayer/files/7101217/Logs-VideoPlayingCorrectly.txt)

**Comment by kim-vde** [comment]
> The content plays fine on all devices, including the FireTVs that that it messes up on, but seemingly out of nowhere the device will start showing this error on a single video in a video loop of 8-15 videos, and the corruption will be there EVERY time the video plays.

Just to be sure I understand, are you testing on the different devices with the *exact same content*?

My understanding is that the same content plays fine on other devices and most of the time on FireTVs, but that it sometimes fails on FireTVs? And that, when it fails on FireTVs, it automatically gets resolved after some time (still for the same content). Is that correct?

Could you clarify what you mean by "video loop"?

> Does anything look odd to you in these logs?

I couldn't find any useful information in there I am afraid. Did you spot any error that was present in Logs-VideoWithStaticError.txt but not in Logs-VideoPlayingCorrectly.txt?

What you could try is to use another decoder for FireTVs by passing a `MediaCodecSelector` to a `DefaultRenderersFactory` that removes the problematic decoder. You can then pass this `RenderersFactory` to the player builder. If this solves the problem, it shows that the decoder is the cause. You can usually find the name of the decoder(s) used in the logs.

**Comment by SAPierce2006** [comment]
> Just to be sure I understand, are you testing on the different devices with the exact same content?

Yes.

> My understanding is that the same content plays fine on other devices and most of the time on FireTVs, but that it sometimes fails on FireTVs? And that, when it fails on FireTVs, it automatically gets resolved after some time (still for the same content). Is that correct?

No. Sometimes it resolves with a restart, but not always. Over the weekend I saw on a FireTV that it happened and then all FireTV applications could no longer play any videos. Youtube, Hulu, Netflix, Amazon Prime, Our app - all of them would freeze the moment they would try to play any video.

This seems to me to point strongly to a decoder bug.

> Could you clarify what you mean by "video loop"?

Sure - Our application plays videos on a loop and never stops. It repeats itself when it gets to the end. Users select what videos play on their video loop.

> What you could try is to use another decoder for FireTVs by passing a MediaCodecSelector to a DefaultRenderersFactory that removes the problematic decoder. You can then pass this RenderersFactory to the player builder. If this solves the problem, it shows that the decoder is the cause. You can usually find the name of the decoder(s) used in the logs.

I was able to replicate your coding steps pretty easily. I made a `FireTvMediaCodecSelector`. The only other codec available is the Google Codec it seems, and it's not hardware accelerated. How big of a deal is that? I'm testing it and I'm seeing some lag in some of our videos. I'm going to investigate whether or not that's because those videos aren't encoded correctly.

**Comment by kim-vde** [comment]
> I was able to replicate your coding steps pretty easily. I made a FireTvMediaCodecSelector. The only other codec available is the Google Codec it seems, and it's not hardware accelerated. How big of a deal is that? 

This means that the decoding will be slower. But if the problem doesn't reproduce with the new decoder, then we can be more confident that this is a bug in the original decoder.

Which decoder was used by default? I will check if we already have bugs reported for this decoder.

**Comment by SAPierce2006** [comment]
> Which decoder was used by default? I will check if we already have bugs reported for this decoder.

OMX.MTK.VIDEO.DECODER.AVC

**Comment by kim-vde** [comment]
Just an idea, but could it be that the device does not support the video resolution (as in #7172)?

I couldn't find a bug describing a similar issue with OMX.MTK.VIDEO.DECODER.AVC.

**Comment by SAPierce2006** [comment]
Is this decoder potentially on other devices? Should I remove it everywhere I see this coder, or only on FireTV?

**Comment by kim-vde** [comment]
> Is this decoder potentially on other devices?

Yes.

> Should I remove it everywhere I see this coder, or only on FireTV?

Everywhere I would say.

Note that it is still possible that the input file is invalid or that the bug comes from FireTV software (and therefore that there is no problem with the decoder). It is a bit suspicious that we never had any similar issue reported (that I am aware of). You should report this to the FireTV team.

**Comment by SAPierce2006** [comment]
I'm using the software decoder `OMX.google.h264.decoder`, and I haven't seen the static / corruption yet, but I'm seeing dropped frames on some videos. Sometimes it stops happening on the videos, other times it continues to repeat. What do you recommend I do to try and resolve this lag with the software decoder?

```
MediaCodecLogger: 19421.SW.video.avc Dropping frame
```
```
I/MediaCodecLogger: 19421.SW.video.avc.TotalVideoPlaybackTimeMs = 102235
I/MediaCodecLogger: 19421.SW.video.avc.videoResolution.FHD = 102235
I/MediaCodecLogger: 19421.SW.video.avc.StreamAvgFramerate = 29.266
I/MediaCodecLogger: my.package.internal.debug.SW.video.avc.StreamFramerate = 29.751
I/MediaCodecLogger: 19421.SW.video.avc.OneConsecutiveFramesDropped = 52
I/MediaCodecLogger: 19421.SW.video.avc.TwoConsecutiveFramesDropped = 97
I/MediaCodecLogger: 19421.SW.video.avc.TotalFramesDropped = 246
I/MediaCodecLogger: 19421.SW.video.avc.MaxConsecutiveFramesDropped = 6
I/MediaCodecLogger: 19421.SW.video.avc.ASAP-P5Ms = 268
I/MediaCodecLogger: 19421.SW.video.avc.ASAP-P2Ms = 19
I/MediaCodecLogger: 19421.SW.video.avc.ASAP-P3Ms = 28
I/MediaCodecLogger: 19421.SW.video.avc.FirstFrameReadyMs = 214
I/MediaCodecLogger: 19421.SW.video.avc.StreamAvgBitrateInKbps = 1744
```

I'm also seeing this in the logs a lot:
```
E/FrameEvents: updateAcquireFence: Did not find frame.
E/FrameEvents: updateAcquireFence: Did not find frame.
E/FrameEvents: updateAcquireFence: Did not find frame.
E/FrameEvents: updateAcquireFence: Did not find frame.
```

**Comment by SAPierce2006** [comment]
> Note that it is still possible that the input file is invalid

Every time we open the app, we have a very slow process running in the background that verifies the checksums of each of the videos.

**Comment by ojw28** [comment]
I would suggest you file an issue on [Amazon's issue tracker](https://github.com/amzn/exoplayer-amazon-port/issues/) to try and get to the bottom of this. Falling back to software decoding is a good diagnostic test, but it's probably not going to be a great solution in production. Particularly if you're playing high resolution content, you'll most likely observe that the software decoder cannot keep up (which you seem to be observing, based on your post above). If you file an issue with Amazon, please post a link to it here for cross-referencing. Thanks!

> Every time we open the app, we have a very slow process running in the background that verifies the checksums of each of the videos.

That only checks the file is the same as the file you have on the server. It doesn't check that the file doesn't violate one of the media specs that it implements (e.g., the MP4 and AVC specs) in some way.

**Comment by SAPierce2006** [comment]
> I would suggest you file an issue on Amazon's issue tracker to try and get to the bottom of this.

I'll do that.

> That only checks the file is the same as the file you have on the server. It doesn't check that the file doesn't violate one of the media specs that it implements (e.g., the MP4 and AVC specs) in some way.

We are encoding our videos using ffmpeg. I kind of implicitly assumed that ffmpeg properly handles encoding. Is that not necessarily the case? What would be the best way to check?

**Comment by ojw28** [comment]
If you're using FFmpeg and not enabling any esoteric features, then you can probably be reasonably confident that it'll be doing the right thing (just because it's widely used and relatively hardened as a result).

**Comment by ScottPierce** [comment]
We've been using 2 commands:
```
ffmpeg -i "<input>" -c:v libx264 -preset slower -profile:v high -level:v 4.0 -c:a aac -b:a 192k -pix_fmt yuv420p "<output>"
```

Because of some solvable corruption we've been seeing on FireTV, we've recently swapped to encode all new videos with
```
ffmpeg -i "<input>" -c:v libx264 -preset faster -crf 23 -c:a aac -b:a 128k -pix_fmt yuv420p "<output>"
```

Is `-profile:v high` an esoteric feature? I've definitely seen more people using CRF, which is why we swapped to it. The reason we actually chose to use profile high instead of crf is because of the firetv video codec support which supports `H.264. Hardware accelerated up to 1080p @ 60fps, High Profile up to Level 4.`
